vllm/engine/llm_engine.py

```diff
@@ -642,81 +646,80 @@ def step(self) -> List[RequestOutput]:
+                gpu_cache_usage=gpu_cache_usage,
+                cpu_cache_usage=cpu_cache_usage,
+                num_prompt_tokens=0,
+                num_generation_tokens=0,
+                time_to_first_tokens=[],
+                time_per_output_tokens=[],
+                time_e2e_requests=[],
+            )
 
-        # Discard the old stats.
-        self.num_prompt_tokens = [(t, n) for t, n in self.num_prompt_tokens
-                                  if now - t < _LOGGING_INTERVAL_SEC]
-        self.num_generation_tokens = [(t, n)
-                                      for t, n in self.num_generation_tokens
-                                      if now - t < _LOGGING_INTERVAL_SEC]
-
-        if len(self.num_prompt_tokens) > 1:
-            total_num_tokens = sum(n for _, n in self.num_prompt_tokens[:-1])
-            window = now - self.num_prompt_tokens[0][0]
-            avg_prompt_throughput = total_num_tokens / window
-        else:
-            avg_prompt_throughput = 0.0
-        if len(self.num_generation_tokens) > 1:
-            total_num_tokens = sum(n
-                                   for _, n in self.num_generation_tokens[:-1])
-            window = now - self.num_generation_tokens[0][0]
-            avg_generation_throughput = total_num_tokens / window
-        else:
-            avg_generation_throughput = 0.0
-
-        total_num_gpu_blocks = self.cache_config.num_gpu_blocks
-        num_free_gpu_blocks = (
-            self.scheduler.block_manager.get_num_free_gpu_blocks())
-        num_used_gpu_blocks = total_num_gpu_blocks - num_free_gpu_blocks
-        gpu_cache_usage = num_used_gpu_blocks / total_num_gpu_blocks
-
-        total_num_cpu_blocks = self.cache_config.num_cpu_blocks
-        if total_num_cpu_blocks > 0:
-            num_free_cpu_blocks = (
-                self.scheduler.block_manager.get_num_free_cpu_blocks())
-            num_used_cpu_blocks = total_num_cpu_blocks - num_free_cpu_blocks
-            cpu_cache_usage = num_used_cpu_blocks / total_num_cpu_blocks
-        else:
-            cpu_cache_usage = 0.0
-
-        record_metrics(
-            avg_prompt_throughput=avg_prompt_throughput,
-            avg_generation_throughput=avg_generation_throughput,
-            scheduler_running=len(self.scheduler.running),
-            scheduler_swapped=len(self.scheduler.swapped),
-            scheduler_waiting=len(self.scheduler.waiting),
+        # Compute IterationStats.
+        prompt_run = scheduler_outputs.prompt_run
+        num_prompt_tokens = scheduler_outputs.num_batched_tokens if prompt_run else 0
+        num_generation_tokens = 0 if prompt_run else scheduler_outputs.num_batched_tokens
+
+        # Get Latency Timings.
+        time_last_iters = []
+        time_e2e_requests = []
+        for seq_group in scheduler_outputs.scheduled_seq_groups:
+            # Time since last token. (n.b. updates seq_group.last_token_time)
+            time_last_iters.append(seq_group.get_last_latency(now))
```

Comment:
Since vllm is generating tokens in batches (groups). All the generations in the same batch will have the same latency. I think it makes sense to log just one, number, how long the last generation batch ran for, like in this pr(#2302).

Reply:
The feature request asked for a histogram (so we can compute P50, P90, P95, P99, and mean on the server side), so we need to log latency for each individual request rather than at the iteration level. Otherwise we will be biasing the percentile calculations to batches with fewer requests.

Additionally, logging per request latency requires each `seq_group` to have its own internal tracker of `last_token_time` so you have to pass `now` to each one to update the internal state.

Reply:
I guess this is ok.